### ISSUE TITLE
New P3 64x32 Issue

### ISSUE THREAD

**Issue by 556duckvader** [2dom/PxMatrix#119]
I am having an issue where not all the LED's are lighting up.  I have followed Brian's guide and I have verified my cabling multiple time and I just can't see why its not working.  Below are some images of the issue.
![IMG_7600](https://user-images.githubusercontent.com/17590678/62843323-10cd2f00-bc87-11e9-9a61-8d0fc9cd5a30.jpg)
![IMG_7601](https://user-images.githubusercontent.com/17590678/62843324-10cd2f00-bc87-11e9-824d-1757fe3b2eee.jpg)

**Comment by 2dom** [comment]
can you post the pattern test, post the result as gif/video and share your code ?

**Comment by 556duckvader** [comment]
Here is the code I used for the Mario Image.  I am working on getting the video of the test pattern uploaded.
[CodeUsed.txt](https://github.com/2dom/PxMatrix/files/3538843/CodeUsed.txt)

**Comment by maditnerd** [comment]
I have a similar issue, 4 lines works and then the next 4 doesn't works.
I also followed Brian's guide (I bought the matrix on aliexpress with the link, he provided on his tetris clock repo)

https://github.com/witnessmenow/WiFi-Tetris-Clock

**Comment by maditnerd** [comment]
I contacted the manufacturer, who confirmed other people has issues with ESP32/ESP8266 and suggested to try it with a Raspberry Pi.

I try the matrix using https://github.com/hzeller/rpi-rgb-led-matrix/ and it works fine with default settings.

Time to find what is going wrong with the ESP32, i'll report what I try here :-)

**Comment by maditnerd** [comment]
It seems related to #104

Chipset on my matrix

* RuiChips RUC7258 (Led Gate Driver) : http://www.ruichips.com/uploads/file/20180816/20180816053859223.pdf
* Chipone ICN2037 (16-Channel Constant Current LED Sink Driver with Dual Latch) http://en.chiponeic.com/content/details36_245.html
* Chipone ICN74HC245 (Logic Control) http://en.chiponeic.com/content/details38_95.html

**Comment by serestre** [comment]
I have same problem

**Comment by serestre** [comment]
This was the response of the provider:

"hello friend, please connect it to raspberry pi controller. ESP8266 controller is updated, not support our led panels recently. Any questions, please feel free to contact me."

**Comment by scarecrow2k6** [comment]
I too have the exact same problem, tried with RPi works great.

**Comment by 2dom** [comment]
Did you try with and without fastUpdate ... might be a timing issue ....

**Comment by 2dom** [comment]
Cannot get hold of that RUC7258 pdf ... does any one of you happen to have it?

**Comment by 2dom** [comment]
@556duckvader I would assume your display is 8 scan. 

Set ` display.begin(8);`

**Comment by serestre** [comment]
This is the Chipset on my matrix

![64X32 P3-6432-2121-16s-d1 0](https://user-images.githubusercontent.com/11726268/64141996-d4957600-cdcf-11e9-8a11-22c7f9e38d87.jpg)

![WIN_20190902_21_43_37_Pro](https://user-images.githubusercontent.com/11726268/64142036-fdb60680-cdcf-11e9-858b-bf0e76d180bd.jpg)
![WIN_20190830_13_45_04_Pro](https://user-images.githubusercontent.com/11726268/64142048-07d80500-cdd0-11e9-9cc4-e46b98396627.jpg)
![WIN_20190902_21_48_43_Pro](https://user-images.githubusercontent.com/11726268/64142062-158d8a80-cdd0-11e9-9f78-83a4d70fd16d.jpg)

**Comment by texynz** [comment]
Same issue, same chipset. Removing E makes no difference, neither does fastupdate. Obviously how the library latches to the next row

**Comment by texynz** [comment]
RUC7258 pdf
http://www.ruichips.com/uploads/file/20180819/20180819083248457.pdf

**Comment by texynz** [comment]
Did some testing and used the following:

  display.drawPixel(0,0,myRED);
  display.drawPixel(0,1,myRED);
  display.drawPixel(0,2,myRED);
  display.drawPixel(0,3,myRED);
  //
  display.drawPixel(0,4,myGREEN);
  display.drawPixel(0,5,myGREEN);
  display.drawPixel(0,6,myGREEN);
  display.drawPixel(0,7,myGREEN);
  //
  display.drawPixel(0,8,myBLUE);
  display.drawPixel(0,9,myBLUE);
  display.drawPixel(0,10,myBLUE);
  display.drawPixel(0,11,myBLUE);


It should give three vertical lines of red, green,blue. But only shows RED and BLUE. If I remove the D pin leave it floating I can see the green show up (flicking) in position 2, but also in position 4 of which I have not programmed. Depending on the state of the "flicker:" if I earth the D, then I can get the position 2 GREEN to stay on. But the other states will now not light up. 
So in conclusion it its leading me to think that on a row change, the D pin needs to be constantly toggled. Funnyily the D pin on the ESP is sitting at 2.5V... and even if I just ground the D pin then I only get the GREEN POSITION.

Will have to have more of a play, but hopefully that info will help someone investigate further

**Comment by 2dom** [comment]
Did you set the right scan in 

display.begin(x)?

**Comment by texynz** [comment]
Yes set it to 16. Tried others also just in case it was some odd setting but no difference. I am running an older pxmatrix.h file that doesn't have an compile issue with the esp8266, but even on the latest build on an esp32 I get the same results.

**Comment by zakrava** [comment]
Hi, I have P3 64x32 1/16 panel https://www.aliexpress.com/item/1615847398.html and it has another issue, when I drive it with esp8266 it turns all leds white or show two white stripes.
It has FM6126A, TC7258EN and 74HC245C chips and pin E not grounded.

**Comment by zakrava** [comment]
Ok, false alarm, I needed just call display.setDriverChip(FM6126A); like it is in mario example for esp32 and now it works like a charm. Sorry and thanks for great library.

**Comment by witnessmenow** [comment]
> 
> 
> I have a similar issue, 4 lines works and then the next 4 doesn't works.
> I also followed Brian's guide (I bought the matrix on aliexpress with the link, he provided on his tetris clock repo)
> 
> https://github.com/witnessmenow/WiFi-Tetris-Clock

Hey @maditnerd , can you confirm where you used the link and I'll at the very least update it with a note to this issue

Sorry that it didn't work, it was the link i used for myself!

**Comment by maditnerd** [comment]
@witnessmenow No problem i'll used it with a Raspberry Pi.
Yes, it is the link from your youtube video : EMvjy3z from **SRYLED Display Store**
Maybe they changed the chipset ? 

I should have tell you first, I was hoping it was simple to fix, but It seems to be more complicated then I would have thought. 

Next time, i'll ask the seller if their matrix is compatible with ESP32. They responded pretty quickly so it's a good idea to do before buying.

**Comment by 2dom** [comment]
Does anyone of you live in central Europe and can send a spare matrix I can test with?

**Comment by egrekov** [comment]
They also bought a panel 64x32 P3, tried a bunch of sketches and everywhere 4 stripes shine, 4 do not shine ...
![rgb1](https://user-images.githubusercontent.com/1834096/65764940-49926c00-e140-11e9-9b82-180119504bed.jpeg)
![rgb2](https://user-images.githubusercontent.com/1834096/65764941-49926c00-e140-11e9-9c53-0e31b839930e.jpeg)

**Comment by witnessmenow** [comment]
I don't have a problem board, but I would be happy to contribute to you getting one from the same seller @2dom

**Comment by 2dom** [comment]
How about this ... tell me the seller and I will get one myself ... if it works you guys can buy me a beer :)

**Comment by witnessmenow** [comment]
I have updated the readme on the WiFi clock with the warning about this display. Does anyone have a good source from Aliexpress at the moment.

**Comment by zakrava** [comment]
Hi, I bought P3 64x32 panel, but there is FM6126A and with esp8266 brightness doesn't work, from here https://www.aliexpress.com/item/1615847398.html
I was looking for other P3 64x32 display, but it looks like it has the same "fm6126a layout" https://www.aliexpress.com/item/32728985432.html

**Comment by texynz** [comment]
https://s.click.aliexpress.com/e/dj9WupK0
This is the panel I bought. The issue is that the chips may change depending on the batch. I didn't get the FM chip and have the issue with the every 4 rows. If you float the a b C pins in you hand like someone suggested the entire screen lights up but flickers. The data is correctly being shifted in but the timing for the scan is either out or requires a different of triggering

**Comment by egrekov** [comment]
Conducted a little research, maybe it will somehow help
![schema](https://user-images.githubusercontent.com/1834096/65812786-c4619280-e1e5-11e9-9c36-2c628a2a00de.jpeg)

**Comment by egrekov** [comment]
Presumably the impulse of line C is supposedly short and because of this OUT4-8 do not have time to grow to the desired amplitude, like A, B, D

**Comment by 2dom** [comment]
Did you try adding a few nops to the code yet?

**Comment by egrekov** [comment]
Unfortunately not so strong as to add changes.

**Comment by 2dom** [comment]
Could you try adding delayMicroseconds(1); to the very end of the setMux function?

**Comment by egrekov** [comment]
I tried to add delayMicroseconds(1); in set_mux. [commit](https://github.com/egrekov/PxMatrix/commit/508da87f44a56ec80d453a225cf9d8e01acc5179)
But I had to use an older version of the library [pxMatrix.zip](https://github.com/2dom/PxMatrix/files/3610615/PxMatrix.zip), otherwise I get an error SPI, how in [#issue129](https://github.com/2dom/PxMatrix/issues/126)

**Comment by 2dom** [comment]
No worries...will wait for the panel to arrive (Brian thankfully donated one) and I will try to figure out what the timing issues are. You scope plots are very helpful ! Thanks!

**Comment by egrekov** [comment]
Add delayMicroseconds(1); only line C [commit](https://github.com/egrekov/PxMatrix/commit/33a8d49bea703b36597d78ac90e3380b2e341c47)
Everything started to display correctly.  There are small artifacts when a large picture is drawn, for example, mario.
![clock](https://user-images.githubusercontent.com/1834096/65835008-9326b680-e2fa-11e9-93e5-88a19fff30d3.jpg)
![circle](https://user-images.githubusercontent.com/1834096/65835014-a46fc300-e2fa-11e9-8803-6bbcdec1365c.jpg)
1. [Youtube video - Mario](https://youtu.be/nwVtCHrKraE)
2. [Youtube video - Banana](https://youtu.be/PccuIavyugk)
3. [Youtube video - Circle](https://youtu.be/3IXsdAyPuL4)
4. [Youtube video - Clock](https://youtu.be/foNBCUxge6I)

**Comment by 2dom** [comment]
Can you try putting the delay after digitalWrite(_C_PIN,HIGH); YOu could also try increasing the delay if that does not do the trick.

**Comment by texynz** [comment]
Nice work. I was so close but had the delay in a different place. If you increase the delay it will reduce the flicker and artifact.

**Comment by 2dom** [comment]
So what are you using now...could you post the snippet of code ?

**Comment by texynz** [comment]
Same place as egrekov in the binary part of setmux for line C but delay of 2us. delay must before the pull high.

**Comment by egrekov** [comment]
By advice @texynz increase delayMicroseconds to 2 and sent pull request to @2dom, to include support for the new display in the library code.

**Comment by egrekov** [comment]
Nevertheless, the addition of microseconds 1-10 in delayMicroseconds, adds a flicker to the display, of course this fixes the problem temporarily, but the picture is not so pleasant for the eye.

**Comment by 2dom** [comment]
Will look into it as soon as I have a panel ...

**Comment by 2dom** [comment]
You can try the latest master. I added a option to control the delay in every stage of the multiplexing.

Try adding `display.setMuxDelay(0,1,0,0,0);` after `display.begin()`.

**Comment by 2dom** [comment]
This would a 1us after the B channel

**Comment by 2dom** [comment]
@witnessmenow Thanks for the panel and the matrix shield - working just fine !

**Comment by scarecrow2k6** [comment]
@2dom Very sorry about this but I can't seem to figure out what is going on. Somewhat of a noob here... tried with both Arduino IDE and VSC- PlatformIO

![Arduino_VSCODE_error](https://user-images.githubusercontent.com/6678214/67170915-973a7700-f382-11e9-90de-cf2850a1c8fc.png)

**Comment by egrekov** [comment]
@scarecrow2k6 You need to update the library
https://github.com/2dom/PxMatrix/blob/master/PxMatrix.h

**Comment by egrekov** [comment]
Spent more testing. it turns out a working commit 4d8c6d4, but I had to comment out lines 654-657 otherwise it did not compile.
`PxMatrix.h:655:123: error: expected ';' before ')' token`
Beginning with the de2df96 commit, problems have already appeared, and at the last commit, not a single LED is working at all.
![not work](https://user-images.githubusercontent.com/1834096/67206525-40d54300-f42b-11e9-8f8c-17da7b5c165b.jpg)

**Comment by 2dom** [comment]
Thanks for the feedback @egrekov . Left a stray return in the function :) Check again with the latest master please!

**Comment by egrekov** [comment]
@2dom, thank you so much, everything works, you're just super)

**Comment by 2dom** [comment]
No worries ... Glad that it works :)

**Comment by zakrava** [comment]
Hi, I have 64x32 P2.5 16S V1.0 panel with ICN2037 and RUC7258 and when I connect it, display is totally blank.
I tried to play with setMuxDelay, but it doesn't help at all. I have latest master.

**Comment by tobobus** [comment]
Maybe this helps. see also issue 139.

Via the "discussions" from the site of Morphing Digital Clock by HariFun i tested a suggestion from "timz3818 " to disable "#define double_buffer " and voila this works.
No black panel-screen after PX 1.7.0 and ESP 8266.
It works for Hari's clock, the remix and further variations.

I have no ideer what double buffer does in those scetches, but disabling works somehow.

**Comment by zakrava** [comment]
I disabled double_buffer a long time ago, otherwise my P3 and P4 panels didn't work.
But P2.5 with two ICN74HC245TS, four RUC7258 and a lot of ICN2037BP is not showing a single pixel.
It is from here https://www.aliexpress.com/item/32650905490.html but is has different chips and layout than on the picture there.

**Comment by icepick7** [comment]
@zakrava
Hello, I have a P3 display witch the exact same ic's an the same behavior, no pixel lights up. Except the pattern_test and the double_buffer Demo works as expected. I cant understand that behavior :(

**Comment by Porokoura** [comment]
> 
> 
> I disabled double_buffer a long time ago, otherwise my P3 and P4 panels didn't work.
> But P2.5 with two ICN74HC245TS, four RUC7258 and a lot of ICN2037BP is not showing a single pixel.
> It is from here https://www.aliexpress.com/item/32650905490.html but is has different chips and layout than on the picture there.

 I currently have 3 different 64 * 32 displays, all with different chip sets. One of them does not have a RUC7258 circuit. This is the only one that works seamlessly with ESP32. There are different problems with those with the RUC725 (eg issue # 152) All 3 of my monitors run smoothly on the Arduino Mega, unfortunately the Arduino's performance is not sufficient for my needs.

**Comment by 2dom** [comment]
Do you use PxMatrix on Arduino Mega?

**Comment by 2dom** [comment]
If so you could try reducing SPI speed

**Comment by Porokoura** [comment]
I don't use PxMatrix in arduino. I just wanted to point out that displays were working.

**Comment by masiorama** [comment]
I'm experiencing the same problem with a panel with the same chips as @maditnerd, and no pixel light at all.

**Comment by DanielRTRD** [comment]
I am in the same boat.

**Comment by Picolo12** [comment]
> Sie kauften auch ein Panel 64x32 P3, probierten ein paar Skizzen aus und überall leuchten 4 Streifen, 4 leuchten nicht ...
> 
> PS. Jetzt hat der Verkäufer bereits gewarnt, dass es Probleme mit dem esp8266 und dem esp32 gibt, aber es gibt keine Probleme mit dem Himbeer-Pi
> 
> ![rgb1](https://user-images.githubusercontent.com/1834096/65764940-49926c00-e140-11e9-9b82-180119504bed.jpeg)
> ![rgb2](https://user-images.githubusercontent.com/1834096/65764941-49926c00-e140-11e9-9c53-0e31b839930e.jpeg)

можно вашу библиотеку? самостоятельно я не смогу 
![62843324-10cd2f00-bc87-11e9-824d-1757fe3b2eee](https://user-images.githubusercontent.com/61686639/77404413-faa55680-6db1-11ea-8f9d-3810f7812e1a.jpg)

**Comment by DanielRTRD** [comment]
Check out #117

**Comment by t3chn0m4g3** [comment]
@DanielRTRD Thank you, this pointed me in the right direction.

I bought a P3 panel (64*32) with the name **E506652 DCHY-M 2013**. While the same panel with **1946** at the end comes delivered with the **FM6124** the **2013** comes with **ICN2037BP**.

The solution for PxMatrix with ESP32 and as such for TetrisClock (shout out to @witnessmenow, thank you!) is to add the following line after `display.begin(16)`.
```
  display.setMuxDelay(1,1,1,1,1);
```

Reducing the speed for just channel B (as the example states in #117) results in flickering for me. Adding delays for all gets rid of flickering in my case.

**Comment by bruno-m-antunes-alb** [comment]
> Hi, I have 64x32 P2.5 16S V1.0 panel with ICN2037 and RUC7258 and when I connect it, display is totally blank. I tried to play with setMuxDelay, but it doesn't help at all. I have latest master.

Did you manage to make this display work? Thanks
